docarray/array/storage/base/backend.py

```diff
@@ -21,6 +23,17 @@ def _init_storage(
         **kwargs,
     ):
         self._load_offset2ids()
+        self._register_exit_hook()
+
+    def _register_exit_hook(self):
+        weakref_del = WeakMethod(self.__del__)
+
+        def exit_hook():
+            if weakref_del() is not None:
+                weakref_del()()
+
+        atexit.register(exit_hook)
+        self._deleted = False
```

Comment:
not a fan of this state? why is it needed?

Reply:
why not handle this in the context manager?

Reply:
Hrmm it's so that it doesn't try to persist the DA twice on SystemExit (program exit).

When SystemExit occurs, the Python Interpreter will first go through all the **hooks** registered by the **atexit** module.
Variables then start getting destroyed which will trigger their **dunder del destructor**.

Using the **dunder del won't work** for our case because the modules our destructors depend on may have already been deallocated by the Python Interpreter at the time the destructor is called in dunder del.
Thus, we want it such that during SystemExit, the **atexit hook** is the method that is used to persist the DA because it will be evaluated at a time when the dependencies are still alive.

But the dunder del will still trigger after the **atexit hook** is complete. So we need a way to make the destructor singleton (it only triggers once, subsequent calls will do nothing).
The proposal at the moment is to implemented the singleton by having this self._deleted flag that tracks whether or not the destructor has been called.

There are of course other ways to implement the singleton destructor behavior so I'm open to suggestions on how to go about this